lib/http/2/client.rb

```diff
@@ -45,19 +52,23 @@ def upgrade
       new_stream(state: :half_closed_local)
     end
 
+    def self.settings_header(**settings)
+      frame = Framer.new.generate(type: :settings, stream: 0, payload: settings)
+      Base64.urlsafe_encode64(frame[9..-1])
+    end
+
+    private
+
     # Emit the connection preface if not yet
     def send_connection_preface
-      return unless @state == :waiting_connection_preface
-      @state = :connected
```

Comment:
Why did you remove the `@state` transition?

Reply:
Because it broke too many specs and it seems like something which doesn't really relate directly to state, but is more like, have we sent it yet? Feel free to dig into it and propose a different solution.